### PR TITLE
IterativeDeepeningAlphaBetaSearch implemented

### PR DESCRIPTION
**Pull request by Manthan-R-Sheth**
Most of the adaptation is from AIMA3e branch. 

**Changes Made** :-
1. ```ActionStore<A>``` inner class is replaced with a ```TreeSet``` to improve the complexity of insertion.
1. Code is cleaned and adjusted according to AIMA4e conventions.

**Path Ahead** :-
1. An interface will be made to increase the functionality, or ```SearchForAdversarialActionFunction``` can be used directly.
1. Tests will be written. 

Since AphaBetaSearch is used with an additional ```depth``` parameter, a class similar to ```DepthLimitedTreeSearch``` could have been used.

### CODE REVIEW

File: extra/src/main/java/aima/extra/search/uninformed/IterativeDeepeningAlphaBetaSearch.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+/**
+ * @author manthan.
+ */
+public class IterativeDeepeningAlphaBetaSearch<A, S, P> implements SearchForAdversarialActionFunction<S, A> {
```
Comment (ctjoreilly) [inline extra/src/main/java/aima/extra/search/uninformed/IterativeDeepeningAlphaBetaSearch.java]: Could you order <A, S, P> the same as AlphaBetaSearch i.e. <S, A, P> and add @param descriptions to the class JavaDoc in a similar way.

**Review by ctjoreilly — CHANGES_REQUESTED** [review]
This looks good, could you address the minor comment I made and also rebase/squash as necessary so I can apply as a patch and try out before merging.

Thanks

Ciaran

**Comment by Manthan-R-Sheth** [comment]
@ctjoreilly 
Updated as said.

File: extra/src/main/java/aima/extra/search/uninformed/IterativeDeepeningAlphaBetaSearch.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+import aima.core.search.api.Game;
+import aima.core.search.api.SearchForAdversarialActionFunction;
+
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+/**
+ * @author manthan.
```
Comment (ctjoreilly) [inline extra/src/main/java/aima/extra/search/uninformed/IterativeDeepeningAlphaBetaSearch.java]: As this is based largely on the AIMA3e version of the code could you add the original author's name here as well (i.e. @author Ruediger Lunde, I would also suggest taking over the description). The convention is that the author who contributed the most to a particular implementation should be listed first. If an author has been involved or code is closely based on a prior implementation, there name should remain included.

File: test/src/test/java/aima/test/unit/search/adversarial/SearchForAdversarialActionFunctionTest.java [hunk 3]
```diff
@@ -28,22 +29,23 @@
...
+	public <S, A, P> A searchForAdversarialAction(Game<S, A, P> game, Integer... params) {
+		SearchForAdversarialActionFunction<S, A> searchFn;
+		if (params != null) {
+			if ("MinimaxDecision".equals(searchFunctionName)) {
+				searchFn = new MinimaxDecision<>(game);
+			} else {
+				searchFn = new AlphaBetaSearch<>(game);
+			}
+		} else {
+			searchFn = new IterativeDeepeningAlphaBetaSearch<>(game, params[0], params[1], params[2]);
```
Comment (ctjoreilly) [inline test/src/test/java/aima/test/unit/search/adversarial/SearchForAdversarialActionFunctionTest.java]: NOTE: IterativeDeepeningAlphaBetaSearch is never tested! At this point params is null and the way the test below is configured AlphaBetaSearch is all that is called and only one test contains the params for the IterativeDeepening calls. This code should be re-configured to just be:

	public <S, A, P> A searchForAdversarialAction(Game<S, A, P> game, Integer... params) {
		SearchForAdversarialActionFunction<S, A> searchFn;
		
		if ("MinimaxDecision".equals(searchFunctionName)) {
			searchFn = new MinimaxDecision<>(game);
		} else if ("AlphaBetaSearch".equals(searchFunctionName)) {
			searchFn = new AlphaBetaSearch<>(game);
		} else {
			searchFn = new IterativeDeepeningAlphaBetaSearch<>(game, params[0], params[1], params[2]);
		}
		
		return searchFn.apply(game.initialState());
	}

File: test/src/test/java/aima/test/unit/search/adversarial/SearchForAdversarialActionFunctionTest.java [hunk 4]
```diff
@@ -111,7 +113,14 @@ public void testAIMA3eFig5_2() {
...
-		
+
 		Assert.assertEquals(new GoAction("B"), searchForAdversarialAction(game));
+
+		Game<InState, GoAction, String> game2 = new BasicGame<>(() -> new InState("A"), aima3eFig5_2PlayerFn,
+				Map2DFunctionFactory.getActionsFunction(aima3eFig5_2),
+				Map2DFunctionFactory.getResultFunction(aima3eFig5_2), aima3eFig5_2TerminalTestPredicate,
+				aima3eFig5_2UtilityFn);
+		Assert.assertEquals(new GoAction("B"), searchForAdversarialAction(game2, 100, 0, 200));
+
```
Comment (ctjoreilly) [inline test/src/test/java/aima/test/unit/search/adversarial/SearchForAdversarialActionFunctionTest.java]: This is just repeating the prior test and adding params (not used currently) specifically for the iterative deepening search. All you need to do is add the params to each of the searchForAdversarialAction calls, as they are only intended to be used by the IterativeDeepening algorithm currently and don't affect the other 2 algorithms. All it needs to be is this:

	@Test
	public void testAIMA3eFig5_2() {
		Game<InState, GoAction, String> game = new BasicGame<>(() -> new InState("A"), aima3eFig5_2PlayerFn,
				Map2DFunctionFactory.getActionsFunction(aima3eFig5_2),
				Map2DFunctionFactory.getResultFunction(aima3eFig5_2), aima3eFig5_2TerminalTestPredicate,
				aima3eFig5_2UtilityFn);
		Assert.assertEquals(new GoAction("B"), searchForAdversarialAction(game, 100, 0, 200));

	}

**Review by ctjoreilly — CHANGES_REQUESTED** [review]
Please rebase/squash as before when you re-submit.

Thanks

Ciaran

**Comment by Manthan-R-Sheth** [comment]
@ctjoreilly 
Ya i realise the mistake of ```IterativeDeepeningAlphaBetaSearch``` never being tested.
Have fixed it though.

**Pull request merged**